### PR TITLE
Fix duckdb leading comments

### LINKED ISSUE #892: adding SQL comments breaks jupysql
From [Slack](https://ploomber-io.slack.com/archives/C028KJCJDRT/p1694819634024439): 

> Did something change with how jupysql handles SQL comments either before or after the %%sql command?  I used to have commented code beforehand which now just errors (fairly certain it didn't before).
If I have commented code one line below the %%sql I get a Success message but no query output (unless I have a something like: %%sql --with first_part --with second_part and then a comment line below.
Thx.

image

### PR DESCRIPTION
**Pull request by marshallwhiteorg**
## Describe your changes
Fix empty result in certain duckdb `SELECT` and `SUMMARIZE` queries with leading comments

## Issue number

Closes #892

## Checklist before requesting a review

- [x] Performed a self-review of my code
- [x] Formatted my code with [`pkgmt format`](https://ploomber-contributing.readthedocs.io/en/latest/contributing/pr.html#linting-formatting)
- [x] Added [tests](https://ploomber-contributing.readthedocs.io/en/latest/contributing/pr.html#testing) (when necessary).
- [x] Added [docstring](https://ploomber-contributing.readthedocs.io/en/latest/contributing/pr.html#documenting-changes-and-new-features) documentation and update the [changelog](https://ploomber-contributing.readthedocs.io/en/latest/contributing/pr.html#changelog) (when needed)

----
:books: Documentation preview :books:: https://jupysql--895.org.readthedocs.build/en/895/

### CODE REVIEW

**Comment by edublancas** [comment]
@marshallwhiteorg any updates here?

**Comment by marshallwhiteorg** [comment]
@edublancas still some issues, will continue in the AM https://github.com/ploomber/jupysql/actions/runs/6320834470/job/17163937282?pr=895

**Comment by edublancas** [comment]
some of these look like matplotlib issues. I think it got upgraded recently and we had to pin the version:

https://github.com/ploomber/jupysql/blob/cf824ca421ec54c321712197ca01c0c53fa9f449/setup.py#L47

**Comment by marshallwhiteorg** [comment]
@tonykploomber @neelasha23

I have all the tests passing except some of the docs are failing to build. My code is using `sqlglot.parse_one` to parse the duckdb queries and that function doesn't accept this type of query:
```
%%sql duckdb:///
INSTALL 'sqlite_scanner';
LOAD 'sqlite_scanner';
```
The error is in here: https://readthedocs.org/projects/jupysql/builds/22054160/

Here's an example.
```
In [44]: from sqlglot import parse_one
In [45]: parse_one("INSTALL 'sqlite_scanner'", dialect="duckdb")
```
```
...
ParseError: Invalid expression / Unexpected token. Line 1, Col: 24.
  INSTALL 'sqlite_scanner'
```

Can you see what I am doing wrong, or do you know why this used to work? Is there some preprocessing step that sqlglot requires for this query? It looks like valid duckdb syntax based on their docs. Also, it works with double quotes but not single quotes. Thanks!

**Comment by neelasha23** [comment]
I can see this query is failing:
```
%%sql
SELECT 
    name, 
    friends[1] AS first_friend, 
    likes.pizza AS likes_pizza, 
    likes.tacos AS likes_tacos
FROM read_json_auto('people.json')
```

The issue might be in this line : ` friends[1] AS first_friend,`

It looks like `sqlglot` had a release yesterday : https://pypi.org/project/sqlglot/#history
You can check the release notes to see if there's any breaking changes. Or you can pin `sqlglot` version <  18.8.0 and rerun the CI. If it fixes the problem open an issue that the latest sqlglot is causing queries to break. @marshallwhiteorg

File: src/sql/connection/connection.py [hunk 1]
```diff
@@ -717,34 +719,40 @@ def _connection_execute(self, query, parameters=None):
...
         if self._requires_manual_commit:
-            # calling connection.commit() when using duckdb-engine will yield
-            # empty results if we commit after a SELECT statement
-            # see: https://github.com/Mause/duckdb_engine/issues/734
-            if is_select and self.dialect == "duckdb":
-                return out
+            # Calling connection.commit() when using duckdb-engine will yield
+            # empty results if we commit after a SELECT or SUMMARIZE statement,
+            # see: https://github.com/Mause/duckdb_engine/issues/734.
+            if self.dialect == "duckdb":
```
Comment (edublancas) [inline src/sql/connection/connection.py:733]: we've had some weird experiences when using sqlglot for parsing (we're actually planning to remove it but haven't done so)

you can use `sqlparse.parse` instead to replicate this functionality. sqlparse won't detect things like `SUMMARIZE df` so you might need to do some manual parsing but it's better than using sqlglot

some quick stuff I tried:

```python
In [1]: import sqlparse

In [2]: sqlparse.parse("select * from stuff")
Out[2]: (<Statement 'select...' at 0x107704200>,)

In [3]: p=sqlparse.parse("select * from stuff")[0]

In [4]: p.get_type()
Out[4]: 'SELECT'

In [5]: sqlparse.parse("select * from stuff")[0].get_type()
Out[5]: 'SELECT'

In [6]: sqlparse.parse("summarize")[0].get_type()
Out[6]: 'UNKNOWN'

In [7]: sqlparse.parse("with x as (select * from y) select * from x")[0].get_type()
Out[7]: 'SELECT'

In [8]: sqlparse.parse("from stuff")[0].get_type()
Out[8]: 'UNKNOWN'

In [9]: sqlparse.parse("from stuff")
Out[9]: (<Statement 'from s...' at 0x1075773E0>,)

In [10]: sqlparse.parse("summarize stuff")
Out[10]: (<Statement 'summar...' at 0x107AFD700>,)

In [11]: p=sqlparse.parse("summarize stuff")[0]

In [12]: p
Out[12]: <Statement 'summar...' at 0x10864FCA0>

In [13]: p.get_real_name()

In [14]: p.token_first()
Out[14]: <Identifier 'summar...' at 0x10864F5A0>

In [15]: token=p.token_first()

In [16]: token
Out[16]: <Identifier 'summar...' at 0x10864F5A0>

In [17]: token.value
Out[17]: 'summarize stuff'

In [18]:

```

File: src/sql/connection/connection.py [hunk 2]
```diff
@@ -717,34 +719,40 @@ def _connection_execute(self, query, parameters=None):
...
+                is_duckdb_sqlalchemy = not self.is_dbapi_connection
+                if is_duckdb_sqlalchemy:
```
Comment (edublancas) [inline src/sql/connection/connection.py:735]: let's also move the parsing logic into a new function and add tests, let's ensure the solution is robust and it can accurately detect the things we want to detect

File: src/tests/test_magic.py [hunk 3]
```diff
@@ -1987,3 +1987,73 @@ def test_accessing_previously_nonexisting_file(ip_empty, tmp_empty, capsys):
...
+        "count": (5,),
+        "null_percentage": ("0.0%",),
+    }
+
+    df = pd.DataFrame(
+        data=dict(
+            memid=[1, 2, 3, 5, 8],
+        ),
+    )
+    _ = df
```
Comment (edublancas) [inline src/tests/test_magic.py:2013]: I guess you added this because of the linting. you can add `# no qa` so the linter doesn't complain because this `_ = df` line is a bit confusing

File: src/tests/test_magic.py [hunk 4]
```diff
@@ -1987,3 +1987,73 @@ def test_accessing_previously_nonexisting_file(ip_empty, tmp_empty, capsys):
...
+
+    ip_empty.run_cell("%sql duckdb://")
+
+    out = ip_empty.run_cell("%sql /* x */ SUMMARIZE df").result
+    assert out.dict() == expected_summarize
+
+    out = ip_empty.run_cell("%sql /*x*//*x*/ SUMMARIZE /*x*/ df").result
+    assert out.dict() == expected_summarize
```
Comment (edublancas) [inline src/tests/test_magic.py:2021]: you can use pytest's parametrize, that's better than having multiple tests in a single function

**Review by edublancas — CHANGES_REQUESTED** [review]

**Comment by marshallwhiteorg** [comment]
@edublancas @idomic @neelasha23 Okay I've made the requested changes, please let me know if you see issues with the new version.

I was skeptical that `sqlparse` would be up to the task but I was only able to make it fail once where `sqlglot` succeeded.
For the query `SELECT FROM table WHERE (column = 'value'` `sqlglot` correctly marks it as invalid due to the missing close paren but `sqlparse` does not. I marked the test as `xfail`. Let me know if you have any more test cases you'd like me to add.

File: src/sql/connection/connection.py [hunk 5]
```diff
@@ -1187,4 +1174,27 @@ def set_sqlalchemy_isolation_level(conn):
...
+def detect_duckdb_summarize_or_select(query):
+    """
+    Checks if the SQL query is a DuckDB SELECT or SUMMARIZE statement.
+
+    Note:
+    Assumes there is only one SQL statement in the query.
+    """
+    statements = sqlparse.parse(query)
+    if statements:
+        assert len(statements) == 1
```
Comment (edublancas) [inline src/sql/connection/connection.py:1186]: this assert statement will break jupysql. if a user tries to run >1 statement it'll fail here

can we support this fix for >1 statements?

File: src/sql/connection/connection.py [hunk 6]
```diff
@@ -1187,4 +1174,27 @@ def set_sqlalchemy_isolation_level(conn):
...
+        stype = statements[0].get_type()
+        if stype == "SELECT":
+            return True
+        elif stype == "UNKNOWN":
+            # Further analysis is required
+            sql_stripped = sqlparse.format(query, strip_comments=True)
+            words = sql_stripped.split()
+            return len(words) > 0 and (
```
Comment (edublancas) [inline src/sql/connection/connection.py:1194]: can you add a test with an empty SQL query? (would `words[0]` break?)

File: src/tests/test_connection.py [hunk 7]
```diff
@@ -1184,3 +1185,55 @@ def test_database_in_directory_that_doesnt_exist(tmp_empty, uri, expected):
         SQLAlchemyConnection(engine=create_engine(uri))
 
     assert expected in str(excinfo.value)
+
+
+query_expected_outputs = [
```
Comment (edublancas) [inline src/tests/test_connection.py:1190]: there's no need to declare a separate variable. you can just include this in the parameterize decorator

File: src/tests/test_connection.py [hunk 8]
```diff
@@ -1184,3 +1185,55 @@ def test_database_in_directory_that_doesnt_exist(tmp_empty, uri, expected):
...
+    ("SELECT column FROM (SELECT * FROM table WHERE column = 'SELECT') AS x", True),
+    # Invalid SQL returns false
+    ("INSERT INTO table (column) VALUES ('SELECT')", False),
+    pytest.param(
+        "SELECT FROM table WHERE (column = 'value'",
+        False,
+        marks=pytest.mark.xfail(
+            reason="sqlparse does not notice the missing close paren"
+        ),
+    ),
```
Comment (edublancas) [inline src/tests/test_connection.py:1221]: we can remove this test case, it's ok. if the SQL is invalid, the SQL engine will just complain and raise an error

**Review by edublancas — CHANGES_REQUESTED** [review]
minor stuff. remember to respond to PR comments with a link to your changes: https://ploomber-contributing.readthedocs.io/en/latest/contributing/responding-pr-review.html

File: src/sql/connection/connection.py (hunk 5 above)
Comment (marshallwhiteorg) [reply src/sql/connection/connection.py:1186]: I do not see a place where this function would be called with >1 statement, the only place it's used is in `_connection_execute` which has this guard against multiple statements: https://github.com/ploomber/jupysql/pull/895/files#diff-62e3d6289e3242b271e6df3a7f201ab003f8ce73a5daf1a824cc2b4d892513d1L721-R722

Can you clarify if I am missing something?

File: src/sql/connection/connection.py (hunk 6 above)
Comment (marshallwhiteorg) [reply src/sql/connection/connection.py:1194]: I have a test case with an empty query https://github.com/ploomber/jupysql/pull/895/files#diff-ecfdf2a1f3adc925d3c32b8aa52540474a10a4160d552f281d7dbd85a0260170R1200

but do you mean that the `len(words) > 0` check is not required?

File: src/tests/test_connection.py (hunk 8 above)
Comment (marshallwhiteorg) [reply src/tests/test_connection.py]: [fixed](https://github.com/ploomber/jupysql/pull/895/files#diff-ecfdf2a1f3adc925d3c32b8aa52540474a10a4160d552f281d7dbd85a0260170R1221)

File: src/tests/test_connection.py (hunk 7 above)
Comment (marshallwhiteorg) [reply src/tests/test_connection.py]: [fixed](https://github.com/ploomber/jupysql/pull/895/files#diff-ecfdf2a1f3adc925d3c32b8aa52540474a10a4160d552f281d7dbd85a0260170R1190)

File: src/sql/connection/connection.py (hunk 5 above)
Comment (edublancas) [reply src/sql/connection/connection.py:1186]: you're right. let's switch the assert for a `NotImpplementedError` then. asserts are non-informative and should only be used in unit tests

File: src/sql/connection/connection.py (hunk 6 above)
Comment (edublancas) [reply src/sql/connection/connection.py:1194]: I mean passing `""` (empty string) as input, would that break the test?

**Comment by edublancas** [comment]
@marshallwhiteorg addressed the comments. please check the open conversations and request my review when done

File: src/sql/connection/connection.py (hunk 6 above)
Comment (marshallwhiteorg) [reply src/sql/connection/connection.py:1194]: The location of it moved slightly but here is the empty string test: https://github.com/ploomber/jupysql/pull/895/files#diff-ecfdf2a1f3adc925d3c32b8aa52540474a10a4160d552f281d7dbd85a0260170R1205

In the repl:
```
In [1]: from sql.connection import detect_duckdb_summarize_or_select

In [2]: detect_duckdb_summarize_or_select("")
Out[2]: False
```

It's because `sqlparse.parse("")` returns an empty collection
```
In [3]: import sqlparse

In [4]: sqlparse.parse("")
Out[4]: ()
```

File: src/sql/connection/connection.py (hunk 5 above)
Comment (marshallwhiteorg) [reply src/sql/connection/connection.py]: [fixed](https://github.com/ploomber/jupysql/pull/895/files#diff-62e3d6289e3242b271e6df3a7f201ab003f8ce73a5daf1a824cc2b4d892513d1R1186-R1187)

**Comment by marshallwhiteorg** [comment]
@edublancas Thanks for the comments, I just responded to both and made the `assert` -> `NotImplementedError` change, see my comment. Ready for review again

**Review by edublancas — DISMISSED** [review]
lgtm. @neelasha23 please run some manual tests and let us know if everything looks good

**Review by neelasha23 — DISMISSED** [review]
lgtm!

**Comment by neelasha23** [comment]
please resolve merge conflicts @marshallwhiteorg

**Comment by edublancas** [comment]
@marshallwhiteorg please resolve merge conflicts and let me know so we can merge this

**Comment by marshallwhiteorg** [comment]
@edublancas @neelasha23 Fixed merge conflicts

**Comment by edublancas** [comment]
@marshallwhiteorg tests are failing

**Comment by marshallwhiteorg** [comment]
@edublancas fixed

**Review by edublancas — APPROVED** [review]
